File: priv/def/test/exdl_joystick.erl
%%
%%  Copyright (c) 2012 David Koch
%%
%%  See the file "license.terms" for information on usage and redistribution
%%  of this file, and for a DISCLAIMER OF ALL WARRANTIES.
%%
%%     $Id: exdl_joystick.erl,v 1.5 2005/06/29 14:45:38 dgud Exp $
%%

-module(exdl_joystick).

-include("exdl.hrl").

%% 1- EXPORT

%-compile(export_all).

-export( [ numJoysticks/0
         , name/1
         , open/1
         , opened/1
         , index/1
         , numAxes/1
         , numBalls/1
         , numHats/1
         , numButtons/1
         , update/0
         , eventState/1
         , getAxis/2
         , getHat/2
         , getBall/4
         , getButton/2
         , close/1
         ]
       ).

%% 2- INCLUDE

-include("exdl_joystick.hrl").

%% 3- IMPORT

-import( exdl_driver
       , [ call/2
         , cast/2
         ]
       ).

%% 4- FUNCTIONS

%% Func:       numJoysticks
%% Args:       none
%% Returns:    Exdl
%% C-API func: int SDL_NumJoysticks(void)

numJoysticks()
 -> case
      call
      ( ?SDL_NumJoysticks
      , [
        ]
      )
    of
      <<Exdl:?Int
      >>
        -> Exdl
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       name
%% Args:       DeviceIndex
%% Returns:    Exdl
%% C-API func: const char* SDL_JoystickName(int device_index)

name(DeviceIndex)
 -> case
      call
      ( ?SDL_JoystickName
      , <<DeviceIndex:?Int
        >>
      )
    of
      0
        -> { nullptr, ?MODULE, ?LINE }
    ; <<Exdl:?String
      >>
        -> { buffptr, Exdl }
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       open
%% Args:       DeviceIndex
%% Returns:    Exdl
%% C-API func: SDL_Joystick* SDL_JoystickOpen(int device_index)

open(DeviceIndex)
 -> case
      call
      ( ?SDL_JoystickOpen
      , <<DeviceIndex:?Int
        >>
      )
    of
      0
        -> { nullptr, ?MODULE, ?LINE }
    ; <<Exdl:?_PTR
      >>
        -> { buffptr, Exdl }
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       opened
%% Args:       DeviceIndex
%% Returns:    Exdl
%% C-API func: int SDL_JoystickOpened(int device_index)

opened(DeviceIndex)
 -> case
      call
      ( ?SDL_JoystickOpened
      , <<DeviceIndex:?Int
        >>
      )
    of
      <<Exdl:?Int
      >>
        -> Exdl
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       index
%% Args:       Joystick
%% Returns:    Exdl
%% C-API func: int SDL_JoystickIndex(SDL_Joystick* joystick)

index(Joystick)
 -> case
      call
      ( ?SDL_JoystickIndex
      , <<Joystick:?_PTR
        >>
      )
    of
      <<Exdl:?Int
      >>
        -> Exdl
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       numAxes
%% Args:       Joystick
%% Returns:    Exdl
%% C-API func: int SDL_JoystickNumAxes(SDL_Joystick* joystick)

numAxes(Joystick)
 -> case
      call
      ( ?SDL_JoystickNumAxes
      , <<Joystick:?_PTR
        >>
      )
    of
      <<Exdl:?Int
      >>
        -> Exdl
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       numBalls
%% Args:       Joystick
%% Returns:    Exdl
%% C-API func: int SDL_JoystickNumBalls(SDL_Joystick* joystick)

numBalls(Joystick)
 -> case
      call
      ( ?SDL_JoystickNumBalls
      , <<Joystick:?_PTR
        >>
      )
    of
      <<Exdl:?Int
      >>
        -> Exdl
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       numHats
%% Args:       Joystick
%% Returns:    Exdl
%% C-API func: int SDL_JoystickNumHats(SDL_Joystick* joystick)

numHats(Joystick)
 -> case
      call
      ( ?SDL_JoystickNumHats
      , <<Joystick:?_PTR
        >>
      )
    of
      <<Exdl:?Int
      >>
        -> Exdl
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       numButtons
%% Args:       Joystick
%% Returns:    Exdl
%% C-API func: int SDL_JoystickNumButtons(SDL_Joystick* joystick)

numButtons(Joystick)
 -> case
      call
      ( ?SDL_JoystickNumButtons
      , <<Joystick:?_PTR
        >>
      )
    of
      <<Exdl:?Int
      >>
        -> Exdl
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       update
%% Args:       none
%% Returns:    ok
%% C-API func: void SDL_JoystickUpdate(void)

update()
 -> cast
    ( ?SDL_JoystickUpdate
    , [
      ]
    )
.

%% Func:       eventState
%% Args:       State
%% Returns:    Exdl
%% C-API func: int SDL_JoystickEventState(int state)

eventState(State)
 -> case
      call
      ( ?SDL_JoystickEventState
      , <<State:?Int
        >>
      )
    of
      <<Exdl:?Int
      >>
        -> Exdl
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       getAxis
%% Args:       Joystick, Axis
%% Returns:    Exdl
%% C-API func: Sint16 SDL_JoystickGetAxis(SDL_Joystick* joystick, int axis)

getAxis(Joystick, Axis)
 -> case
      call
      ( ?SDL_JoystickGetAxis
      , <<Joystick:?_PTR
        , Axis:?Int
        >>
      )
    of
      <<Exdl:?Sint16
      >>
        -> Exdl
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       getHat
%% Args:       Joystick, Hat
%% Returns:    Exdl
%% C-API func: Uint8 SDL_JoystickGetHat(SDL_Joystick* joystick, int hat)

getHat(Joystick, Hat)
 -> case
      call
      ( ?SDL_JoystickGetHat
      , <<Joystick:?_PTR
        , Hat:?Int
        >>
      )
    of
      <<Exdl:?Uint8
      >>
        -> Exdl
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       getBall
%% Args:       Joystick, Ball, Dx, Dy
%% Returns:    Exdl
%% C-API func: int SDL_JoystickGetBall(SDL_Joystick* joystick, int ball, int* dx, int* dy)

getBall(Joystick, Ball, Dx, Dy)
 -> case
      call
      ( ?SDL_JoystickGetBall
      , <<Joystick:?_PTR
        , Ball:?Int
        , Dx:?_PTR
        , Dy:?_PTR
        >>
      )
    of
      <<Exdl:?Int
      >>
        -> Exdl
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       getButton
%% Args:       Joystick, Button
%% Returns:    Exdl
%% C-API func: Uint8 SDL_JoystickGetButton(SDL_Joystick* joystick, int button)

getButton(Joystick, Button)
 -> case
      call
      ( ?SDL_JoystickGetButton
      , <<Joystick:?_PTR
        , Button:?Int
        >>
      )
    of
      <<Exdl:?Uint8
      >>
        -> Exdl
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       close
%% Args:       Joystick
%% Returns:    ok
%% C-API func: void SDL_JoystickClose(SDL_Joystick* joystick)

close(Joystick)
 -> cast
    ( ?SDL_JoystickClose
    , <<Joystick:?_PTR
      >>
    )
.

File: priv/def/test/exdl_joystick.c
/*
 *  Copyright (c) 2012 David Koch
 *
 *  See the file "license.terms" for information on usage and redistribution
 *  of this file, and for a DISCLAIMER OF ALL WARRANTIES.
 *
 *     $Id: exdl_joystick.c,v 1.5 2005/06/29 14:45:38 dgud Exp $
 */

#include "exdl.h"

EXDL_API(ejoy_numJoysticks)
{
  /* int SDL_NumJoysticks(void) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  CATCH_RESULT(int)

    SDL_NumJoysticks
    (
    );

  THROW_RESULT(int)
}

EXDL_API(ejoy_name)
{
  /* const char* SDL_JoystickName(int device_index) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(int, device_index)
  STACK_FMT_END()

  CATCH_RESULT(const char*)

    SDL_JoystickName
    ( STACK_GET_DATA(device_index)
    );

  THROW_STRING()
}

EXDL_API(ejoy_open)
{
  /* SDL_Joystick* SDL_JoystickOpen(int device_index) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(int, device_index)
  STACK_FMT_END()

  CATCH_RESULT(SDL_Joystick*)

    SDL_JoystickOpen
    ( STACK_GET_DATA(device_index)
    );

  THROW_POINTER()
}

EXDL_API(ejoy_opened)
{
  /* int SDL_JoystickOpened(int device_index) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(int, device_index)
  STACK_FMT_END()

  CATCH_RESULT(int)

    SDL_JoystickOpened
    ( STACK_GET_DATA(device_index)
    );

  THROW_RESULT(int)
}

EXDL_API(ejoy_index)
{
  /* int SDL_JoystickIndex(SDL_Joystick* joystick) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(SDL_Joystick, joystick[])
  STACK_FMT_END()

  CATCH_RESULT(int)

    SDL_JoystickIndex
    ( STACK_GET_ADDR(joystick)
    );

  THROW_RESULT(int)
}

EXDL_API(ejoy_numAxes)
{
  /* int SDL_JoystickNumAxes(SDL_Joystick* joystick) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(SDL_Joystick, joystick[])
  STACK_FMT_END()

  CATCH_RESULT(int)

    SDL_JoystickNumAxes
    ( STACK_GET_ADDR(joystick)
    );

  THROW_RESULT(int)
}

EXDL_API(ejoy_numBalls)
{
  /* int SDL_JoystickNumBalls(SDL_Joystick* joystick) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(SDL_Joystick, joystick[])
  STACK_FMT_END()

  CATCH_RESULT(int)

    SDL_JoystickNumBalls
    ( STACK_GET_ADDR(joystick)
    );

  THROW_RESULT(int)
}

EXDL_API(ejoy_numHats)
{
  /* int SDL_JoystickNumHats(SDL_Joystick* joystick) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(SDL_Joystick, joystick[])
  STACK_FMT_END()

  CATCH_RESULT(int)

    SDL_JoystickNumHats
    ( STACK_GET_ADDR(joystick)
    );

  THROW_RESULT(int)
}

EXDL_API(ejoy_numButtons)
{
  /* int SDL_JoystickNumButtons(SDL_Joystick* joystick) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(SDL_Joystick, joystick[])
  STACK_FMT_END()

  CATCH_RESULT(int)

    SDL_JoystickNumButtons
    ( STACK_GET_ADDR(joystick)
    );

  THROW_RESULT(int)
}

EXDL_API(ejoy_update)
{
  /* void SDL_JoystickUpdate(void) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  SDL_JoystickUpdate
  (
  );
}

EXDL_API(ejoy_eventState)
{
  /* int SDL_JoystickEventState(int state) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(int, state)
  STACK_FMT_END()

  CATCH_RESULT(int)

    SDL_JoystickEventState
    ( STACK_GET_DATA(state)
    );

  THROW_RESULT(int)
}

EXDL_API(ejoy_getAxis)
{
  /* Sint16 SDL_JoystickGetAxis(SDL_Joystick* joystick, int axis) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_LPTR(SDL_Joystick*, joystick)
    STACK_FMT_DATA(int,           axis)
  STACK_FMT_END()

  CATCH_RESULT(Sint16)

    SDL_JoystickGetAxis
    ( STACK_GET_DATA(joystick)
    , STACK_GET_DATA(axis)
    );

  THROW_RESULT(Sint16)
}

EXDL_API(ejoy_getHat)
{
  /* Uint8 SDL_JoystickGetHat(SDL_Joystick* joystick, int hat) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_LPTR(SDL_Joystick*, joystick)
    STACK_FMT_DATA(int,           hat)
  STACK_FMT_END()

  CATCH_RESULT(Uint8)

    SDL_JoystickGetHat
    ( STACK_GET_DATA(joystick)
    , STACK_GET_DATA(hat)
    );

  THROW_RESULT(Uint8)
}

EXDL_API(ejoy_getBall)
{
  /* int SDL_JoystickGetBall(SDL_Joystick* joystick, int ball, int* dx, int* dy) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_LPTR(SDL_Joystick*, joystick)
    STACK_FMT_DATA(int,           ball)
    STACK_FMT_LPTR(int*,          dx)
    STACK_FMT_LPTR(int*,          dy)
  STACK_FMT_END()

  FRAME_FMT_BEGIN()
    FRAME_FMT_DATA(int, exdl)
  FRAME_FMT_END()

  FRAME_SET_DATA(exdl) =

    SDL_JoystickGetBall
    ( STACK_GET_DATA(joystick)
    , STACK_GET_DATA(ball)
    , STACK_GET_DATA(dx)
    , STACK_GET_DATA(dy)
    );

  THROW_FRAME()
}

EXDL_API(ejoy_getButton)
{
  /* Uint8 SDL_JoystickGetButton(SDL_Joystick* joystick, int button) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_LPTR(SDL_Joystick*, joystick)
    STACK_FMT_DATA(int,           button)
  STACK_FMT_END()

  CATCH_RESULT(Uint8)

    SDL_JoystickGetButton
    ( STACK_GET_DATA(joystick)
    , STACK_GET_DATA(button)
    );

  THROW_RESULT(Uint8)
}

EXDL_API(ejoy_close)
{
  /* void SDL_JoystickClose(SDL_Joystick* joystick) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(SDL_Joystick, joystick[])
  STACK_FMT_END()

    SDL_JoystickClose
    ( STACK_GET_ADDR(joystick)
    );
}

